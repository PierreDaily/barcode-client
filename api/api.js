import axios from "axios";
import { Tokens } from "./class";

const serverAddress =
  process.env["SERVER_ADDRESS"] === "development"
    ? "192.168.0.183"
    : "pierredaily.com";
const serverPort = process.env["NODE_ENV"] === "development" ? 3001 : 3000;
const sheme = process.env["NODE_ENV"] === "development" ? "http" : "https";
const serverURL = `${sheme}://${serverAddress}:${serverPort}`;
const auth = new Tokens();

const API = {
  _getNewAccessToken: async function(refreshToken) {
    try {
      return await API.post("/auth", {}, { refresh_token: refreshToken }).then(
        ({ data: { access_token } }) => access_token
      );
    } catch (err) {
      return null;
    }
  },

  _fetch: async (method, path, params, headers) => {
    const options = {
      headers: {
        Authorization: `Bearer ${auth.accessToken}`,
        "Content-Type": "application/json",
        ...headers
      },
      method,
      data: params
    };

    try {
      return await axios(`${serverURL}${path}`, options);
    } catch (err) {
      if (err && err.response && err.response.status === 401) {
        auth.accessToken = await API._getNewAccessToken(auth.refreshToken);

        if (auth.accessToken) {
          const updatedOptions = {
            ...options,
            headers: {
              ...options.headers,
              Authorization: `Bearer ${auth.accessToken}`
            }
          };
          return axios(`${serverURL}${path}`, updatedOptions);
        } else {
          throw new Error("can't get a new access token");
        }
      } else {
        throw err;
      }
    }
  },
  get: (path, headers) => API._fetch("GET", path, null, headers),
  post: (path, params, headers) => API._fetch("POST", path, params, headers),
  put: (path, params, headers) => API._fetch("PUT", path, params, headers)
};

export { API };
