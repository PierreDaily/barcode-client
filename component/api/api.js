import axios from "axios";

const serverAddress =
  process.env["NODE_ENV"] === "development" ? "192.168.0.183" : "5.135.186.3";
const serverPort = 3000;
const sheme = "http";
const serverURL = `${sheme}://${serverAddress}:${serverPort}`;

const API = {
  get: (path, headers) => {
    const options = {
      ...headers,
      "Content-Type": "application/json"
    };
    return axios.get(`${serverURL}${path}`, options);
  },

  post: (path, params, headers) => {
    const options = {
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        ...headers
      },
      method: "POST",
      data: params
    };

    return axios(`${serverURL}${path}`, options);
  },

  put: (path, params, headers) => {
    const options = {
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        ...headers
      },
      method: "PUT",
      data: params
    };

    return axios(`${serverURL}${path}`, options);
  }
};

export default API;
