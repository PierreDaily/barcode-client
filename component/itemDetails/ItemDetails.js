import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { Formik } from "formik";
import * as yup from "yup";
import { Text, View, StyleSheet, Picker } from "react-native";
import { Button, TextInput } from "react-native-paper";
import API from "../api/api";
import logger from "./../../logger";

const getBrandList = () => API.get(`/brand`);

const yupSchema = yup.object().shape({
  brand: yup.string().required(),
  name: yup.string().required()
});

const ItemDetails = ({ route, navigation }) => {
  const [brand, setBrand] = useState(null);
  const [brandList, setBrandList] = useState([]);

  useEffect(() => {
    getBrandList()
      .then(({ data }) => {
        setBrand(data[0].id);
        setBrandList(data);
      })
      .catch(err => {
        logger("brand list can't be fetched");
        alert(err);
      });
  }, []);

  return (
    <View>
      <Text>Barcode: {route.params.barcode}</Text>
      {brand ? (
        <Formik
          initialValues={{ name: "" }}
          validateOnChange={false}
          validationSchema={yupSchema}
          onSubmit={values =>
            navigation.navigate("Item-photo", {
              ...route.params,
              brand: values.brand,
              name: values.name
            })
          }
          testID="form"
        >
          {({ errors, handleSubmit, setFieldValue, values }) => (
            <View>
              <TextInput
                error={errors.name}
                label="Name"
                onChangeText={text => setFieldValue("name", text)}
                value={values.name}
              />

              <Text>Brand</Text>
              <Picker
                selectedValue={values.brand}
                style={[
                  { height: 50, width: 200 },
                  errors.brand && styles.error
                ]}
                onValueChange={itemValue => setFieldValue("brand", itemValue)}
              >
                {brandList.length > 0 &&
                  brandList.map(brand => (
                    <Picker.Item
                      label={brand.name}
                      key={brand.id}
                      value={brand.id}
                    />
                  ))}
              </Picker>
              <Button mode="contained" onPress={handleSubmit}>
                Submit
              </Button>
            </View>
          )}
        </Formik>
      ) : (
        <Text testID="loading">Loading...</Text>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  error: {
    color: "red"
  }
});

ItemDetails.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired
  }),
  route: PropTypes.shape({ params: PropTypes.object })
};

export default ItemDetails;
