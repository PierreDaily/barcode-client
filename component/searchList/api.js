import API from "../api/api";

const customAPI = {
  postBrand: name => {
    return API.post("/brand", { name }).then(({ data }) => data);
  },
  searchBrand: name =>
    name !== ""
      ? API.get(
          `/brand-search/${name}`,
          { name: name.toLowerCase() },
          { "Content-Type": "application/json" }
        ).then(({ data }) => data)
      : Promise.resolve([])
};

export default customAPI;
