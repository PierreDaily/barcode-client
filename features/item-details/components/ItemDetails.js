import React from "react";
import PropTypes from "prop-types";
import { Formik } from "formik";
import * as yup from "yup";
import { View } from "react-native";
import { TextInput } from "react-native-paper";
import { Button } from "../../ui/components";
import styles from "../styles";
import Logo from "../../../assets/img/green-barcode.svg";

const yupSchema = yup.object().shape({
  brandId: yup.number().required(),
  brandName: yup.string().required(),
  name: yup.string().required()
});

const ItemDetails = ({ route, navigation }) => {
  const {
    params: { brandId, brandName, barcode, type }
  } = route;

  return (
    <Formik
      initialValues={{
        name: ""
      }}
      validateOnChange={false}
      validate={async values => {
        const errors = {};
        try {
          await yupSchema.validate({
            brandId,
            brandName,
            name: values.name
          });
        } catch (err) {
          errors[err.path] = err.message;
        }

        return errors;
      }}
      onSubmit={values =>
        navigation.navigate("Item-photo", {
          barcode,
          brand: brandId,
          name: values.name,
          type
        })
      }
      testID="form"
    >
      {({ errors, handleSubmit, setFieldValue, values }) => (
        <View style={styles.container}>
          <Logo style={styles.logo} />
          <View>
            <TextInput
              error={errors.name}
              label="Product name"
              onChangeText={text => setFieldValue("name", text)}
              style={styles.nameInput}
              testID="textInput"
              value={values.name}
            />

            <TextInput
              error={errors.brandName}
              label="Brand Name"
              style={styles.brandInput}
              value={brandName || ""}
              onFocus={() =>
                navigation.navigate("Search", {
                  barcode,
                  type
                })
              }
              testID="search"
            />
          </View>
          <Button
            mode="contained"
            onPress={handleSubmit}
            style={styles.btn}
            testID="submit"
          >
            Next
          </Button>
        </View>
      )}
    </Formik>
  );
};

ItemDetails.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired
  }),
  route: PropTypes.shape({ params: PropTypes.object })
};

export { ItemDetails };
