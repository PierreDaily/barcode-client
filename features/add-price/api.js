import { API } from "../../api";

const customAPI = {
  postPrice: ({ branchId, discountIndex, itemId, totalPrice }) => {
    return API.post("/price", {
      branchId,
      discountIndex,
      itemId,
      totalPrice: totalPrice * 100
    }).then(({ data }) => data);
  }
};

export default customAPI;
