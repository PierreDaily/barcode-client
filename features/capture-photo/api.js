import { API } from "../../api";

const customAPI = {
  postItem: ({ barcode, barcodeType, brand, name, imgUri }) => {
    let data = new FormData();
    data.append("image", {
      uri: imgUri,
      name: "image.jpg",
      type: "image/jpeg"
    });
    data.append("barcode", barcode);
    data.append("barcode_type", barcodeType);
    data.append("name", name);
    data.append("brand", brand);

    return API.post(`/item`, data, {
      "content-type": `multipart/form-data`
    }).then(({ data }) => data);
  }
};

export default customAPI;
