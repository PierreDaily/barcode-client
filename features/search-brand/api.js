import { API } from "../../api";

const customAPI = {
  postBrand: name => {
    return API.post("/brand", { name }).then(({ data }) => data);
  },
  searchBrand: name =>
    name !== ""
      ? API.get(`/brand-search/${name}`).then(({ data }) => data)
      : Promise.resolve([])
};

export default customAPI;
