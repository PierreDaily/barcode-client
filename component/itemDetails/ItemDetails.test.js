import React from "react";
import {
  act,
  render,
  fireEvent,
  waitForElement,
  flushMicrotasksQueue
} from "react-native-testing-library";
import ItemDetails from "./ItemDetails";
import api from "../api/api";
jest.mock("../api/api");
import renderer from "react-test-renderer";

let props;

beforeEach(() => {
  props = {
    navigation: { navigate: jest.fn() },
    route: { params: { barcode: "123456789" } }
  };
});

test("renders correctly", () => {
  const tree = renderer.create(<ItemDetails {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("useEffect API call works", async () => {
  api.get.mockResolvedValue({
    data: [
      {
        id: 1,
        name: "nestle"
      },
      {
        id: 2,
        name: "coca cola"
      }
    ]
  });
  const { getByTestId, getAllByTestId } = render(<ItemDetails {...props} />);

  expect(getAllByTestId("loading")).toHaveLength(1);
  await act(async () => flushMicrotasksQueue());
  expect(getByTestId("form")).toBeTruthy();
  expect(api.get).toHaveBeenCalledWith("/brand");
});
