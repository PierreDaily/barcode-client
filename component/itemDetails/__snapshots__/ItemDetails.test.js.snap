// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <Text>
    Barcode: 
    123456789
  </Text>
  <Text
    testID="loading"
  >
    Loading...
  </Text>
</View>
`;
